### ISSUE TITLE
Inference Redux

### ISSUE THREAD

**Issue by acl-cqc** [CQCL/hugr#639]
First a couple of notes:
* `main_loop` should not need to iterate more than once. It claims `solve_metas` can add new edges to the equality graph, but only if it finds `Constraint::Equals`; a single iteration of merge_equal_metas should remove all Constraint::Equals and thus all edges from the equality graph
* There is only ever one variable - the input-extensions for the Hugr root node (right?)

So, I propose removing/splitting main_loop, and the two calls to main_loop, as follows:
1. Solve equality constraints (merge_equal_metas). Maybe we only generate the equality graph (by scanning constraints) just prior to this. The result is a load of entries in `shunted`, new entries in `constraints` (i.e. keyed by the merged vars, but containing only Plus's), and a smaller set of variables (i.e. `remaining` in `main_loop` - the new vars, and everything that wasn't shunted.)
   * Since `shunted` is a member of `UnificationContext` I suspect that remaining should be too. It'd be nice to use `constraints.keys()` but not sure whether that works(?)
2. Solve cycles of Plus constraints (for variables in `remaining`). Probably build that graph now. Variables on cycles can be shunted/merged; note they had no Equals constraints, but may still have Plus constraints pointing out of the cycle.
   * Since Petgraphs's `tarjan_sccs` allows to return SCCs in an order (where cycle A depends upon cycle B, then B before A, say), we *could* solve the remaining constraints as we go here, but lets not.
   * Instead we merely add more entries to `shunted`, make some more metas and move remaining non-cyclic `constraints` around, and build a new map `lower_bound` (from merged meta to all Extensions on the cycle it was merged from).
   * In particular, the case of a single-node cycle on itself must be handled here (whether we delete the constraint and add a `lower_bounds`, or "merge" the single node into a new meta without the cycle, doesn't really matter).
3. Note that *nothing so far has depended upon the graph variables (root node input extensions)*.... And *nothing after this generates new variables or alters `constraints`*, so there's a clear split here....
4. For the single graph variable, we could
   * Require Hugr root node to have specified input extensions
   * Take root node input ExtensionSet as an extra parameter
   * If root node is open, solve it to it's `lower_bound`, or empty ExtensionSet otherwise.
5. Now we solve. The graph of Plus constraints is acyclic, so first error if any leaf (with no outgoing Plus constraints) doesn't have a solution. Then, proceed in topological sort order (leaves first) ==> at each node, there are some outgoing Plus constraints, all of which are solved. Compute solution for each of those, error if different, then
   * in the current Hugr-with-lift-nodes, check the solution includes the `lower_bound`
   * in proposed Hugr-without-lift-nodes (below), add the `lower_bound`.

Hugr-without-Lift-nodes ==> update validation to require extensions at source of edge are a subset of extensions of target of edge. Beyond change in step 5, I think that's all that's required.

**Comment by acl-cqc** [comment]
> 4. For the single graph variable, we could

Alternatively, if there really is only one variable, if we add to the `lower_bound` the total of all Constraint::Plus delta's on the path from each meta to that variable, then we have a partial solution: the "complete" "solution" is just obtained by taking union of partial-solutions for each meta with the value of the variable, which could be left to the caller.

**Comment by croyzor** [comment]
I'm not really sure that there should be only one "variable" in a hugr. What if the hugr is a Module containing 3 dataflow graphs - then the root of each of those graphs should be a variable?

**Comment by acl-cqc** [comment]
> I'm not really sure that there should be only one "variable" in a hugr. What if the hugr is a Module containing 3 dataflow graphs - then the root of each of those graphs should be a variable?

Yes, that's a fair question - seems like a reasonable thing to do (I note merely that it's not what we do at the moment: https://github.com/CQCL/hugr/blob/b3062e9b8d0f081c1ce65295e30066aa16c39ee4/src/extension/infer.rs#L270-L273)

So I think we can still handle items 1+2 first, without any consideration of variable(s), and then handle even multiple variables in stage 5 (e.g. passing an extra `Map<Location,ExtensionSet>` into the topsort), without changing the overall approach. Moreover, I think we can do even that in a few stages / incremental changes.

I'm still thinking I'd *like* to change handling of DFGs such that the `Input` node always has *empty* ExtensionSet; this guarantees computation of correct "delta", whereas current approach means the delta could underapproximate (by anything in input-extensions). The top-level FunctionDefn's in the Module would then have *unused* input-extensions, which TBH seems pretty reasonable. *However* there is a big downside in that the `Constraint::Plus` then has to take two Metas (rather than a Meta and an ExtensionSet) - but I'm nervous as to whether the algorithm above would still work if we changed Constraint in that way.

**Comment by acl-cqc** [comment]
So I think we can break this up into two themes: firstly, reordering steps in the algorithm; secondly, changing the approach to variables. The second is less certain and needs more thought, so on the first...

Following #591 I think all the parts are pretty much there, but the current flow is
1. `main_loop`, which performs multiple iterations of merging equalities and solving constraints;
2. `instantiate_variables` which deals with variables and cycles of plus constraints
3.  a second pass of `main_loop`

I think what we can do is
1. Merge equalities (do not solve constraints) - a single iteration (CCs)
2. Merge plus constraints - a single iteration (SCCs)
3. Apply variables (or not), and solve constraints - a single iteration (topsort)

The latter raises questions (allows to address issues) like, what if my meta `m` is plus-dependent on both a variable `v` and a solvable meta `m2` (e.g. which has a fixed set of (input_)extensions specified in the Hugr) - if we don't have a value for `v`, we can avoid solving `m`; or we can solve `m`, and optionally report the constraint that places on the variable `v`. (E.g. if `m` has `Constraint::Plus({E1,E2}, V)` then the value of `v` must be a subset of `m` and a superset of `m \ {E1,E2}`.) This might want be part of a bigger rethink of variables, however.

**Comment by acl-cqc** [comment]
Wrt. variables, we should probably do #657 first. Then we can explicitly declare variables of kind TypeParam::Extensions, rather than the somewhat implicit "open with no constraints" that we use in inference now.

This would allow returning an ExtensionSolution (that is complete, i.e. defined for all nodes) *containing* those variables, where necessary; and giving the highly-desirable property that instantiating the variables (i.e. substituting them with concrete ExtensionSets) and *then* performing inference, would produce the same answers as performing inference and *then* substituting the variables to concrete ExtensionSets.

**Comment by acl-cqc** [comment]
However, if there are fixed/provided annotations already, then it is unlikely that we can provide a solution (a set of annotations) that is valid for *all* values of any variables (even by expressing the solution in terms of those variables). For example, consider a DFG Hugr with open (variable) input extensions and fixed delta D, where the Input node is wired directly to the Output. The only solution here is where the "variable" takes the value D ! Or, supposing we were able to ask the inference algorithm to infer the delta (#640), but manually specified/forced the extensions for the Output node to D (in which case in the absence of a Lift node the delta must be empty) - again the variable would have to be equal to D.

Hence, it seems we may have to return both
1. A solution (mapping from nodes to extensionsets, for nodes which did not have extensionsets already specified, which may mention variables among the ExtensionIDs) - or else we write this into the Hugr, i.e. overwriting `None`s (but only Nones)
2. Lower- and upper- bounds on the variables (sometimes coinciding) between which the solution is correct. (Note that at present `TypeParam::Extensions` allows declaring a variable that ranges over arbitrary ExtensionSets without any bounds.)

**Comment by acl-cqc** [comment]
It also seems useful to look at solving `Plus` constraints backwards, as it turns out this relates quite closely to the notion of variables:

1. **Chain forwards** (total solutions forwards):
A = Plus(E1, B)
B = Plus(E2, C)

given an exact solution for C, we can compute an exact solution for B, and then A.

2. **Merge** (total solution backwards, partial solution forwards):
A = Plus(E1, B)
C = Plus(E2, B)

Given an exact solution for A, we can partially-solve the first constraint *backwards* and compute that $B \subseteq A \wedge B \supseteq A \setminus E1$. That is, any value in this range works.

Given an inexact solution for B, if we solve the second constraint *forwards* we can see that $C \subseteq A \setminus E2 \wedge C \supseteq A \setminus (E1 \cup E2)$ but this is not sufficient condition for a solution/instantiation to be valid - we need to record the dependency upon B, thus reifying the meta B (/those locations) to a "variable"....

3. **Chain backwards** (partial solution backwards)

In the first example if we have a solution for A, solve A=Plus(E1, B) backwards to get $B \subseteq A \wedge B \supseteq A \setminus E1$, then try to partially-solve B = Plus(E2,C) *backwards*, we again have to reify B, but then can report that `C \subseteq B \wedge C \supseteq B \setminus E2$.

**Comment by acl-cqc** [comment]
This means the three-phase algorithm is 1. merge equalities, 2. merge cycles of plus constraints (deducing lower bounds), to give a DAG of Plus Contraints; and 3. solve the DAG. The last stage could take a bool parameter as to whether to deduce only definite/fixed extension-annotations (i.e. where there is only one possibility), as the existing code does, or whether to also try to deduce/choose *some acceptable* value for other annotations, i.e. that makes the Hugr valid (even if other values might too). I think the latter corresponds roughly to, do we solve Plus constraints backwards; and if so, a reasonable policy might be to prefer the *smallest* extensionset that works.

Even this is not entirely straightforward. E.g. given a single constraint `A: Plus(E1, B)` and a concrete solution for A (that must therefore include E1), the *smallest* solution is `B = A - E1`. (But `B = A` is also valid!). However, if we also have `B: Plus(E1, C)`, and still only a concrete solution for A, then the only solution for B is now `B = A` but there are two choices for `C`.

**Comment by acl-cqc** [comment]
<strike>In more detail
Input: DAG of metas linked only by Plus constraints, and a `lower_bounds: Map<Meta, ExtensionSet>`
1. For each node A in topsort order, i.e. s.t. we process A after B if A has constraint Plus(exts, B). This pass generates either a solution for A or an entry in `lower_bounds`.
* If A has no constraints, set lower_bounds[A] to empty if A does not already have an entry
* Else, A has constraints. Every node B on which A depends has *either* a solution, *or* an entry in lower_bounds.
   * If any B has a solution, compute solution for A; check same solution produced from all solved B. Check for all unsolved B that (lower-bound of B plus extensions added by constraint) is contained within the solution (else, **fail with error**).
   * if no B has a solution, i.e. all B's have only lower-bounds, compute the union over all B's of (lower-bound of B plus extensions added by constraint), union any existing `lower_bounds` entry for A; update `lower_bounds` for A with this.

2. For each node A in reverse order, i.e. s.t. we process A after any B if B has constraint Plus(exts, A). This pass generates a solution for A.
* Consider Plus constraints upon A (sources of these have already been processed and have solutions).
* If A already has a solution, continue. (Maybe check the Plus constraints on A are satisfied).
* Otherwise, take union (over B's that have constraint upon A) of (the solution to B *minus* delta of the Plus constraint). Also union `lower_bounds` of A. This is the new (least-possible) solution for A. By construction from previous step this should work...??</strike>

Say we have
A: Plus(E1, C)
B: Plus(E2, C)

In the absence of any solutions specified already, we compute `lower_bounds` of A={E1}, B={E2}. These then become solutions for A, B, and so C becomes empty-set.

However, supposing we have `A = {E1, E3}` already, then we would need C to be at least `{E3}` and thus `{B}` to be at least `{E2, E3}`. You can imagine a "zig zag" chain - two columns, with all Plus-edges in one direction, and then a solution at one end of the chain. This sounds like it has to propagate all the way down, forwards then backwards then forwards then backwards. So that sounds like a topsort-based solution cannot work.

**Comment by acl-cqc** [comment]
This might be enacted in stages as follows:
1. Add more rigorous notion of variables. There are a few possibilities:
   * DeBruijn-indexed a là "type variables". Thus, we implement a `substitute` function on Hugrs paralleling that on Type(Arg/Param)s. There are a couple of sub-possibilities, but either way we may need to do inference on FuncDefn's in some kind of tree-traversal order rather than all together...
      a. with FuncDefn/cl's able to refer to variables declared outside (as per current #692). This means that when inference infers a variable should be placed into an ExtensionSet on a particular node, we need to know what DeBruijn-index the variable has *there*, i.e. how many intervening binders they have; ick.
      b. with "flat" FuncDefn/cl's that don't refer to enclosing variables. Thus, the `call`s to these defn/cl's must pass any type-variables from enclosing scope. `substitute` gets simpler, and the same variable has the same name everywhere. To avoid confusion I think we would probably have to insist FuncDefn's and their Input-node children have empty input-extensions, but that might also be a good idea in (a) above.
   * Some other kind of ExtensionId identifier. Then we need a different `substitute` that takes a Map<ExtensionId,ExtensionSet> - and we'll probably need a DeBruijn one later (for Tierkreis) *as well* :-(. However, this might make some things much simpler.
2. Remove existing idea of "variable instantiation", and separate algorithm into the three stages - roughly the same code we have ATM, but without mapping things to empty set. We would need a big test-case reorg here to distinguish (a) those that really *want* to do something with variables (in which case use mechanism from 1. above) vs (b) those that actually want the empty ExtensionSet as input; this might just mean changing the root node from open() to pure().
3. Add inference of smallest-possible set by solving Plus's backwards. This would probably "fix" most of the test cases that had to be changed in 2.... (so might we want to skip step 2 and do 2+3 together?)

**Comment by croyzor** [comment]
I'm in agreement. We just need to plan the order in which we do things. I think:
* Split up static edges into "Const" and "Func ref"
* Change the handling of variables in inference to coincide with the treatment of extensionset variables from polymorphic hugrs, and add some test cases of inference on such polymorphic hugrs
* Restructure the code, per the first post in this thread
* Start solving plus constraints backwards

**Comment by acl-cqc** [comment]
Splitting EdgeKind::Static into Const/Func is #695 or another? I'm not sure that's necessarily the right solution to the problem

**Comment by acl-cqc** [comment]
We probably also need to handle type-variables's in ExtensionSets even *without* using these for extension-inference's own notion of variables; I *think* this means (at the least?!) that inference mustn't propagate such a TV outside its binder.

**Comment by croyzor** [comment]
> We probably also need to handle type-variables's in ExtensionSets even _without_ using these for extension-inference's own notion of variables

Agreed!

> I _think_ this means (at the least?!) that inference mustn't propagate such a TV outside its binder.

This is more of a head-scratcher. I feel like if we get to this point, it says there's something scarily non-parametric about the function!

**Comment by acl-cqc** [comment]
> I think if we get to this point, it says there's something scarily non-parametric about the function!

Yes!! ;-). So at the least, we should verify that any variables in any `input_extensions` sets are "in scope" (in `validate_subtree`, where we check TypeArgs for certain leaf ops are too). That'd catch anything apart from some weird awkward case where the variable had "leaked" out *into another place that happened to have an identical variable* (!).

However, if we restrict the input-extensions to empty for FuncDefn/cl, then the output-extensions are necessarily empty too (as delta is zero), and if Consts are monomorphic (as at present), then I think that should prevent any leakage; the Static edge has PolyFuncType that would bind the variable, and any `call` would instantiate that with some ExtensionSet from outside, so we're all good.

Polymorphic lambdas/constants is another issue (#709), but if it's a Const containing a lambda then that's empty delta, so we say the input-extensions and output-extensions of the const are empty, and again anything bound by / inside the PolyFuncType on the static edge should be fine.
